Add include_terminal param

hyanwong opened the pull request:
Fixes #783

jeromekelleher commented on c/tskit/trees.c:
```diff
@@ -4371,7 +4371,8 @@ tsk_tree_map_mutations(tsk_tree_t *self, int8_t *genotypes,
 int TSK_WARN_UNUSED
-tsk_diff_iter_init(tsk_diff_iter_t *self, tsk_treeseq_t *tree_sequence)
+tsk_diff_iter_init(
+    tsk_diff_iter_t *self, tsk_treeseq_t *tree_sequence, bool include_terminal)
```
Let's make this a ``tsk_flags_t options`` argument - this is the approach taken elsewhere in the API, as it's more flexible and breaks less code when new options are added.

jeromekelleher commented on python/tests/test_highlevel.py:
```diff
@@ -655,6 +655,19 @@ def test_edge_diffs(self):
+        for ts in get_example_tree_sequences():
+            edges = set()
+            i = 0
+            for _, e_out, e_in in ts.edge_diffs(include_terminal=True):
```
We need to test the value of left and right here as well. What are they?

jeromekelleher commented on python/tskit/trees.py:
```diff
@@ -3457,15 +3457,17 @@ def edge_diffs(self):
+            report a final removal of all remaining edges), so the number
+            of iterations will be equal to the number of trees in the tree
+            sequence. If True, there is one more iteration, with the last
+            ``edges_out`` value reporting all the edges contained in the final
```
What's be value of left and right here?

hyanwong replied:
Good point. By default it is `right==left==sequence_length`, which seems reasonable to me, so I've documented this and added to to the high_level tests.

jeromekelleher commented on c/tskit/trees.c:
```diff
@@ -4499,10 +4503,10 @@ tsk_diff_iter_next(tsk_diff_iter_t *self, double *ret_left, double *ret_right,
-        if (self->removal_index < self->num_edges) {
+        if (self->removal_index < (tsk_id_t) self->num_edges) {
             right = TSK_MIN(right, edges->right[removal_order[self->removal_index]]);
         }
```
We need some C tests of this to make sure we're not touching any uninitialised memory. Easiest thing to do is to run through the iterator setting a ``parent`` array as usual, and then check at the end that the parent array is all -1 if TSK_INCLUDE_TERMINAL has been set.

hyanwong replied:
Just done this in https://github.com/tskit-dev/tskit/pull/787/commits/b4a3f1d9d26c44b8330bb9bf5276ee7e7c621093 - I've not written many C tests before, so let me know if anything is wrong / unconventional.

jeromekelleher replied:
Nope, these look ideal, thanks (modulo issues @benjeffery points out)

jeromekelleher reviewed (changes requested):
Looks good, but needs more tests. Definitely on the C side we want to exercise this on a few different topology types - what happens on an empty ts? An ordinary ts with several trees? We have to think through all the corner cases.

hyanwong commented:
> Looks good, but needs more tests. Definitely on the C side we want to exercise this on a few different topology types - what happens on an empty ts? An ordinary ts with several trees? We have to think through all the corner cases.

Can we not test an empty ts, multi-tree ts etc at the python level? We have a function `get_example_tree_sequences()` in test_highlevel.py which can return various pathological tree sequences like this. Actually, having looked at it, I see it doesn't test a ts with no edges, so I'll open a PR for that.

jeromekelleher commented:
> Can we not test an empty ts, multi-tree ts etc at the python level? We have a function get_example_tree_sequences() in test_highlevel.py which can return various pathological tree sequences like this. 

No, because we can't run valgrind when Python is running and would therefore not detect memory access issues (which are non-deterministic and differ across platforms, so you can't just say "it doesn't segfault, so it's fine").

There's loads of tests checking for empty tree sequences across the suite - the test_highlevel.py approach isn't how we're doing things any more.

hyanwong commented:
> No, because we can't run valgrind when Python is running and would therefore not detect memory access issues (which are non-deterministic and differ across platforms, so you can't just say "it doesn't segfault, so it's fine").

Ah, OK. Apart from your suggestion above for all `parent[j] == -1`, I'm not familiar with the way you might check memory issues here. But now that I have implemented that, perhaps I can just add `test_multiroot_diff_iter` and `test_empty_diff_iter` functions to test_trees.c?

> There's loads of tests checking for empty tree sequences across the suite - the test_highlevel.py approach isn't how we're doing things any more.

Added anyway: I think there's no harm, and a function to produce a whole set of different semi-pathological tree sequences seems a useful thing to have around.

jeromekelleher commented:
> Ah, OK. Apart from your suggestion above for all parent[j] == -1, I'm not familiar with the way you might check memory issues here. But now that I have implemented that, perhaps I can just add test_multiroot_diff_iter and test_empty_diff_iter functions to test_trees.c?

Once the tests are implemented valgrind is run on them automatically, checking for these errors.

benjeffery commented on python/_tskitmodule.c:
```diff
@@ -9869,7 +9874,8 @@ TreeDiffIterator_init(TreeDiffIterator *self, PyObject *args, PyObject *kwds)
     err = tsk_diff_iter_init(self->tree_diff_iterator,
-            self->tree_sequence->tree_sequence);
+            self->tree_sequence->tree_sequence,
+            include_terminal);
```
Shouldn't this be `options`?

hyanwong replied:
Oh, very good catch. Thanks.

benjeffery commented on c/tests/test_trees.c:
```diff
@@ -380,76 +380,94 @@ verify_tree_next_prev(tsk_treeseq_t *ts)
-            parent[record->edge.child] = record->edge.parent;
-        }
-        /* Now check against the sparse tree iterator. */
+    for (include_terminal = 0; include_terminal <= 1; include_terminal++) {
```
For functions with options I've been using the pattern at https://github.com/tskit-dev/tskit/blob/master/c/tests/test_tables.c#L782 as I think it is clearer what's going on.

hyanwong replied:
Sorry - is that the right link @benjeffery - I can't see anything about options there?

benjeffery replied:
Yeah, that's the right link - basically call the test function with each option. In the link:
```
    test_edge_table_with_options(0);
    test_edge_table_with_options(TSK_NO_METADATA);
```

hyanwong replied:
Oh, I see. Sorry - I was being dense. I agree, much clearer.

benjeffery reviewed (changes requested):
Think I found am incorrect variable usage in the python C layer, and one comment about test style.

jeromekelleher commented on c/tests/testlib.c:
```diff
@@ -48,7 +48,21 @@ const char *single_tree_ex_mutations
       "2    3     1   -1\n";
 
-/* Example from the PLOS paper */
+/*** Example from the PLOS paper ***/
```
Fantastic! Thanks for putting these in, that's a major bonus.

hyanwong replied:
No probs. I just need to sort out the C linting (not set up on my system) and Ben's last comment, and this should be done, ready for 0.3.1.

benjeffery replied:
I can c lint if needed.

hyanwong replied:
Thanks @benjeffery - it would save me quite a lot of hassle if you could lint the C, so please do if that's OK. I've just pushed the commit with the "options" parameter, which is much better (thanks for the tip). I can also squash the commits into 1, or I guess you can, it it's easy (the first one, f9a4a972d346862c5ba22e4673b856a149256c01) is a separate PR, though, so perhaps leave that)

benjeffery replied:
Done, force pushed.

hyanwong replied:
Thanks a lot, Ben!

benjeffery reviewed (changes requested):
Looking good, thanks @hyanwong! We just need a changelog before merge.

hyanwong commented:
> Looking good, thanks @hyanwong! We just need a changelog before merge.

Done, thanks. Note that this includes the change in https://github.com/tskit-dev/tskit/pull/788 as the first commit in the PR, so that can be PR can be closed if this is merged.

jeromekelleher commented:
Let's hold off on this until our bugfix releases have been shipped.

hyanwong commented:
I think this is ready to merge @benjeffery ?

benjeffery reviewed (approved):
LGTM, lets merge.

benjeffery commented:
@hyanwong I've just force pushed to resolve the conflict and add the change to the C changelog too as this is a change to both APIs.
